src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,484 @@
+
+// A tool class for GradientMutator
+// Visit the forward bindings and generate the backward bindings
+class BackwardBindingGenerator : public ExprVisitor {
+ public:
+  /*!
+   * \brief Generate the backward bindings for the corresponding GradientMutator
+   *
+   * \param forward_block The forward dataflow block
+   * \param require_grads The relax variables whose adjoints are needed.
+   * \param post The expression with rewritten inputs.
+   */
+  Expr Generate(const BlockBuilder& builder, const DataflowBlock& forward_block,
+                const Array<Var>& require_grads, const Var& target_var) {
+    this->builder_ = builder;
+
+    // initialize the adjoint of target_var as ones op
+    adjoint_msg_map_.Set(target_var, BuildOnesAdjoint(target_var));
+
+    // We do reverse-mode ad, so visit bindings backwards
+    for (auto it = forward_block->bindings.rbegin(); it != forward_block->bindings.rend(); ++it) {
+      this->VisitBinding(*it);
+    }
+
+    return this->Epilogue(require_grads, target_var);
+  }
+
+  void VisitBinding(const Binding& binding) override {
+    // TODO(chaofan, yixin): support other types of bindings
+    CHECK(binding->IsInstance<VarBindingNode>()) << "now only support VarBindingNode";
+    auto var_binding = binding.as<VarBindingNode>();
+
+    if (adjoint_msg_map_.count(var_binding->var) == 0) {
+      // This var is not used in the bindings handled earlier
+      return;
+    }
+
+    // Meet the definition of binding->var
+    // Create the adjoint var and bind the adjoint value to it
+    EmitAdjoint(var_binding->var, adjoint_msg_map_[var_binding->var], true);
+
+    Expr value = var_binding->value;
+    // TODO(chaofan, yixin): support other types of binding values
+    CHECK(value->IsInstance<CallNode>() || value->IsInstance<TupleNode>() ||
+          value->IsInstance<TupleGetItemNode>() || value->IsInstance<VarNode>() ||
+          value->IsInstance<ConstantNode>())
+        << "now does not support the type of binding value: " << value;
+
+    ExprVisitor::VisitBinding_(var_binding);
+  }
+
+  // Handle the adjoint expr of the inputs of binding
+  // For call node, we would call the registered gradient functions
+  void VisitBinding_(const VarBindingNode* binding, const CallNode* call) override {
+    static const OpAttrMap<FPrimalGradient>& gradient_op_map =
+        Op::GetAttrMap<FPrimalGradient>("FPrimalGradient");
+
+    Var adjoint_var = adjoint_var_map_[binding->var];
+    const Op& call_op = GetRef<Op>(call->op.as<OpNode>());
+    const Array<Expr>& partials =
+        gradient_op_map[call_op](binding->var, GetRef<Call>(call), adjoint_var, builder_);
+    ICHECK(partials.size() == call->args.size()) << "partials number != inputs number";
+
+    for (size_t i = 0; i < partials.size(); ++i) {
+      Expr partial = partials[i];
+      if (!partial->struct_info_.defined()) {
+        UpdateStructInfo(partial, GetStructInfo(call->args[i]));
+      }
+      UpdateAdjoint(call->args[i], partial);
```

Comment:
A caveat when propagating msgs to the arguments of calls is that not all arguments are of tensor type (they might be tuple of ints/shape), which means they don't have well-defined msgs

Reply:
In this case, the gradient function should return `UnknownGradient` to this arguments, and the AD side can handle it. This can be done in the future since now the arguments of our operators are all tensor type.

Reply:
I think we could add tests for it in the future when we support ops like `take`.

Reply:
Sounds good